From c7cccbd862b910e3e18741b692a753fa81f37fc3 Mon Sep 17 00:00:00 2001
From: "Voleti, Murali" <murali.voleti@intel.com>
Date: Wed, 18 Jul 2018 17:43:14 +0200
Subject: [PATCH] Merge pull request #79 in SW_UGW/linux from
 deliver_05.04.02.37_to_ugw_8.x to master

* commit '6ca5edc90374f1372eb11c55a68d6897390ad578':
  WLANRTSYS-2686 - Deliver Wave CV 05.04.02.37
---
 net/mac80211/key.c     | 16 ++++++++++++++++
 net/wireless/nl80211.c |  6 ++----
 2 files changed, 18 insertions(+), 4 deletions(-)

diff --git a/net/mac80211/key.c b/net/mac80211/key.c
index 4c625a325ce2..4df9f6514950 100644
--- a/net/mac80211/key.c
+++ b/net/mac80211/key.c
@@ -677,6 +677,21 @@ int ieee80211_key_link(struct ieee80211_key *key,
 	key->sdata = sdata;
 	key->sta = sta;
 
+	/*
+	 * Silently accept key re-installation without really installing the
+	 * new version of the key to avoid nonce reuse or replay issues.
+	 */
+	if (old_key && key->conf.keylen == old_key->conf.keylen &&
+		!memcmp(key->conf.key, old_key->conf.key, key->conf.keylen)) {
+		ieee80211_key_free_unused(key);
+		ret = 0;
+		goto out;
+	}
+
+	key->local = sdata->local;
+	key->sdata = sdata;
+	key->sta = sta;
+
 	increment_tailroom_need_count(sdata);
 
 	ieee80211_key_replace(sdata, sta, pairwise, old_key, key);
@@ -739,6 +754,7 @@ void ieee80211_enable_keys(struct ieee80211_sub_if_data *sdata)
 		ieee80211_key_enable_hw_accel(key);
 	}
 
+out:
 	mutex_unlock(&sdata->local->key_mtx);
 }
 
diff --git a/net/wireless/nl80211.c b/net/wireless/nl80211.c
index 0ddaea2cbe50..021da0ff2b8d 100644
--- a/net/wireless/nl80211.c
+++ b/net/wireless/nl80211.c
@@ -4889,8 +4889,8 @@ static int nl80211_new_station(struct sk_buff *skb, struct genl_info *info)
 	    !info->attrs[NL80211_ATTR_PEER_AID])
 		return -EINVAL;
 
-	if (!info->attrs[NL80211_ATTR_STA_RSSI])
-		return -EINVAL;
+	if (info->attrs[NL80211_ATTR_STA_RSSI])
+		params.rssi = nla_get_u32(info->attrs[NL80211_ATTR_STA_RSSI]);
 
 	mac_addr = nla_data(info->attrs[NL80211_ATTR_MAC]);
 	params.supported_rates =
@@ -4929,8 +4929,6 @@ static int nl80211_new_station(struct sk_buff *skb, struct genl_info *info)
 	if (!params.aid || params.aid > IEEE80211_MAX_AID)
 		return -EINVAL;
 
-	params.rssi = nla_get_u32(info->attrs[NL80211_ATTR_STA_RSSI]);
-
 	if (info->attrs[NL80211_ATTR_STA_CAPABILITY]) {
 		params.capability =
 			nla_get_u16(info->attrs[NL80211_ATTR_STA_CAPABILITY]);
